runtime/src/accounts.rs

```diff
@@ -278,26 +275,49 @@ impl Accounts {
                             })
                             .unwrap_or_default();
 
-                        if account.executable() && bpf_loader_upgradeable::check_id(account.owner())
+                        if account.executable()
+                            && demote_program_write_locks
+                            && message.is_writable(i, demote_program_write_locks)
+                            && !is_upgradeable_loader_present(message)
                         {
-                            // The upgradeable loader requires the derived ProgramData account
-                            if let Ok(UpgradeableLoaderState::Program {
-                                programdata_address,
-                            }) = account.state()
-                            {
-                                if let Some(account) = self
-                                    .accounts_db
-                                    .load_with_fixed_root(ancestors, &programdata_address)
-                                    .map(|(account, _)| account)
+                            error_counters.invalid_account_index += 1;
+                            return Err(TransactionError::InvalidAccountIndex);
```

Comment:
Would be helpful to return a new error here that better represents the failure

Reply:
Yeah, I went back and forth on that, and ultimately went for less churn. But I'll add a unique error and error_counter